### ISSUE TITLE
Add package `tokenizers`

### ISSUE THREAD

**Issue by josephrocca** [pyodide/pyodide#2816]
There has recently been some exciting progress on bringing Hugging Face's Tokenizers lib to wasm/browsers:

 * @Narsil: https://github.com/huggingface/tokenizers/pull/1009
 * @mbrunel: https://github.com/mithril-security/tokenizers-wasm
 * Discussion: 
   * https://github.com/huggingface/tokenizers/issues/935
   * https://github.com/huggingface/tokenizers/issues/63
   * https://github.com/huggingface/tokenizers/issues/1010

I think it'd be great if the full Python bindings were available in the browser (including JupyterLite) via Pyodide, and it doesn't seem like too much extra work.

@messense has been super helpful over on [this issue](https://github.com/huggingface/tokenizers/issues/1010) and the instructions below now produce a `.whl` that seems to *almost* work in the browser (problem is explained below).

### Build instructions:
Visit [this branch](https://github.com/josephrocca/tokenizers/tree/pyodide) and start a Github Codespace on it, then run these commands:
```bash
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
source $HOME/.cargo/env
rustup toolchain install nightly
rustup target add --toolchain nightly wasm32-unknown-emscripten
rustup component add rust-src --toolchain nightly-x86_64-unknown-linux-gnu
cargo install --git https://github.com/PyO3/maturin.git maturin # tested and working at commit hash 05169de
git clone https://github.com/emscripten-core/emsdk.git # tested and working at commit hash 517e02f
cd emsdk
./emsdk install latest
./emsdk activate latest
source ./emsdk_env.sh
cd ../bindings/python
RUSTUP_TOOLCHAIN=nightly maturin build --release -o dist --target wasm32-unknown-emscripten -i python3.10
```

### The problem
This produces a `.whl` file that seems to work, except for a threading issue as you can see [in this demo](https://josephrocca.github.io/tokenizers-pyodide/demo/) ([source repo](https://github.com/josephrocca/tokenizers-pyodide)). @ryanking13 mentioned in [this comment](https://github.com/pyodide/pyodide/discussions/2720#discussioncomment-3008755) that Pyodide has some problems with threads that come up at runtime, so perhaps that's what's happening here?

For quick reference, here's the error that you'll see in the browser console in the above-linked demo:

```
pyo3_runtime.PanicException: The global thread pool has not been initialized.: ThreadPoolBuildError { kind: IOError(Os { code: 6, kind: WouldBlock, message: "Resource temporarily unavailable" }) }
```

![image](https://user-images.githubusercontent.com/1167575/176789979-a429ce5d-4205-4a23-9588-565799eed5a7.png)

**Comment by ryanking13** [comment]
Thanks for the detailed report @josephrocca.

> ```
> pyo3_runtime.PanicException: The global thread pool has not been initialized.: ThreadPoolBuildError { kind: IOError(Os { code: 6, kind: WouldBlock, message: "Resource temporarily unavailable" }) }
> ```

Yes I think this is related to the runtime threading issue. Maybe Hood have some ideas as he has more experiences on Rust than me.

**Comment by rth** [comment]
Could you try setting [`TOKENIZERS_PARALLELISM=0`](https://github.com/huggingface/tokenizers/blob/28cd3dce2a75d106572392194ff2564574c33235/tokenizers/src/utils/parallelism.rs#L9) env variable? Generally yes, threading would need to be disabled in tokenizers.

**Comment by josephrocca** [comment]
@rth Nice! That solved it!

Demo now works: https://josephrocca.github.io/tokenizers-pyodide/demo/

I've added [another patch](https://github.com/josephrocca/tokenizers/commit/72d6b1d80190b44c6a243afd9ad2d7e6ea5a6bb0) that disables parallelism by default.

Given the custom build code that's in my original post, what would be the best way to get this working with `meta.yaml` so it can go into official Pyodide package index? Is there a way to override the whole build process with a custom script?

Also, I don't think Narsil's wasm updates have made it into a release yet - am I able to source a Github commit hash in `meta.yaml` somehow, and then add patches for that, or would we need to wait for the next release? Looks like releases have been happening every month or two, so might be worth just waiting?

**Comment by julien-c** [comment]
That's so cool!

**Comment by bhugueney** [comment]
Any update on this issue ?
I, for one, would have loved to be able to use tokenizers from JupyterLite.

**Comment by ryanking13** [comment]
@bhugueney 

> Any update on this issue ?

No. But, if you are interested,  I think you can try to build tokenizers yourself and report us what happens. The pyodide build system has been improved a lot since this issue was opened.

**Comment by agriyakhetarpal** [comment]
Reading the thread above, `tokenizers` does not use `pyodide-build`, rather, it uses `maturin`. Given that the latter has been used to build for `wasm32-unknown-emscripten` for packages like Polars, I think it is at a usable state and has been for a reasonable amount of time now. A patch that sets the `TOKENIZERS_PARALLELISM` environment variable to `0` under `#[cfg(target_family = "wasm")]` (or an upstream PR), with a recipe to build with that patch, should be all that is required to build `tokenizers`.

**Comment by YoanSallami** [comment]
Hello, I'm also very interested in running tokenizers in pyodide, is there any progress on it? Thank you :)

**Comment by ryanking13** [comment]
@YoanSallami There was a approach to build tokenizers before, but no progress since then (https://github.com/pyodide/pyodide/pull/3422). PR welcome.
